src/core/execute.c

```diff
@@ -1905,31 +1905,39 @@ static int setup_private_users(uid_t uid, gid_t gid) {
          * we however lack after opening the user namespace. To work around this we fork() a temporary child process,
          * which waits for the parent to create the new user namespace while staying in the original namespace. The
          * child then writes the UID mapping, under full privileges. The parent waits for the child to finish and
-         * continues execution normally. */
+         * continues execution normally.
+         * For unprivileged users (i.e. without capabilities), the root to root mapping is excluded. As such, it
+         * does not need CAP_SETUID to write the single line mapping to itself. */
 
-        if (uid != 0 && uid_is_valid(uid)) {
+        /* Can only set up multiple mappings with CAP_SETUID. */
+        if (have_effective_cap(CAP_SETUID) && uid != 0 && uid_is_valid(uid)) {
                 r = asprintf(&uid_map,
                              "0 0 1\n"                      /* Map root → root */
                              UID_FMT " " UID_FMT " 1\n",    /* Map $UID → $UID */
                              uid, uid);
                 if (r < 0)
                         return -ENOMEM;
         } else {
-                uid_map = strdup("0 0 1\n");            /* The case where the above is the same */
-                if (!uid_map)
+                r = asprintf(&uid_map,
+                             UID_FMT " " UID_FMT " 1\n",
```

Comment:
nitpick: These two blocks now look eerily similar, except for `uid` vs `geteuid()`.

Can I suggest something like this instead?

```c
if (!(have_effective_cap(CAP_SETUID) && uid != 0 && uid_is_valid(uid)))
        uid = geteuid();  /* Map $EUID -> $EUID instead. */
```

And then a single `asprintf` and error checking returning `-ENOMEM` for each case.

Perhaps tweak the logic so it's more readable (use `||`?) And maybe expand the comments to explain the code better, since it's more terse now...

Same for `gid`.